presto-main/src/main/java/com/facebook/presto/operator/Driver.java

```diff
@@ -230,6 +234,11 @@ private void processNewSources()
 
         requireNonNull(duration, "duration is null");
 
+        // if the driver is blocked we don't need to continue
+        if (!previousDriverBlockedFuture.isDone()) {
+            return previousDriverBlockedFuture;
```

Comment:
> This future is already guarded by `exclusiveLock`

I think I don't understand. `processFor(Duration)` is a public (unsynchronized) method that consults  `previousDriverBlockedFuture` before doing any other work. It seems to me that we can e.g. return different content of `previousDriverBlockedFuture` field that we consulted just a line ago, since this can be changed in the meantime. Are you positive it is sufficiently synchronized?
(I am pretty sure `Driver#processFor` isn't going to be called from multiple threads at the same time.. But I'd rather understand that I do not have to make this assumption when reading code of this class.)